src/mozanalysis/experiment.py

```diff
@@ -0,0 +1,509 @@
+        Ref: https://bugzilla.mozilla.org/show_bug.cgi?id=1536644
+
+        Args:
+            spark: The spark context.
+            study_type (str): One of the following strings:
+                - 'pref_flip'
+                - 'addon'
+            end_date (str, optional): Ignore enrollments after this
+                date: for faster queries on stale experiments. If you
+                set `num_dates_enrollment` then do not set this; at best
+                it would be redundant, at worst it's contradictory.
+
+        Returns:
+            A Spark DataFrame of enrollment data. One row per
+            enrollment. Columns:
+                - client_id (str)
+                - enrollment_date (str): e.g. '20190329'
+                - branch (str)
+        """
+        if study_type == 'pref_flip':
+            enrollments = self._get_enrollments_view_normandy(spark)
+
+        elif study_type == 'addon':
+            enrollments = self._get_enrollments_view_addon(spark, self.addon_version)
+
+        # elif study_type == 'glean':
+        #     raise NotImplementedError
+
+        else:
+            raise ValueError("Unrecognized study_type: {}".format(study_type))
+
+        enrollments = enrollments.filter(
+            enrollments.enrollment_date >= self.start_date
+        ).filter(
+            enrollments.experiment_slug == self.experiment_slug
+        )
+
+        if self.num_dates_enrollment is not None:
+            assert end_date is None, "Possibly contradictory instructions!"
+            enrollments = enrollments.filter(
+                enrollments.enrollment_date <= self._get_scheduled_max_enrollment_date()
+            )
+        elif end_date is not None:
+            enrollments = enrollments.filter(
+                enrollments.enrollment_date <= end_date
+            )
+
+        enrollments.cache()
+
+        return enrollments
+
+    def get_per_client_data(
+        self, enrollments, df, metric_list, last_date_full_data, conv_window_start_days,
+        conv_window_length_days, keep_client_id=False
+    ):
+        """Return a DataFrame containing per-client metric values.
+
+        Args:
+            enrollments: A spark DataFrame of enrollments, like the one
+                returned by `self.get_enrollments()`.
+            df: A spark DataFrame containing the data needed to calculate
+                the metrics. Could be `main_summary` or `clients_daily`.
+                _Don't_ use `experiments`; as of 2019/04/02 it drops data
+                collected after people self-unenroll, so unenrolling users
+                will appear to churn. Must have at least the following
+                columns:
+                    - client_id (str)
+                    - submission_date_s3 (str)
+                    - data columns referred to in `metric_list`
+                Ideally also has:
+                    - experiments (map)
+            metric_list: A list of columns that aggregate and compute
```

Comment:
Can you note that these metrics will be applied over groups of (client, branch) data?

Reply:
Good point - done!